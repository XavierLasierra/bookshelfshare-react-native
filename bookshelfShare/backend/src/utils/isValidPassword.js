const bcrypt = require("bcrypt");

function isValidPassword(password) {
  return bcrypt.compareSync(password, this.password);
}

module.exports = { isValidPassword };
