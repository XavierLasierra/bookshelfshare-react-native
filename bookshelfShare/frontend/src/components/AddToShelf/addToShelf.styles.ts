import {StyleSheet} from 'react-native';
import stylesConstants from '../../styles/styles.constants';

export default StyleSheet.create({
  addToShelfTitle: {
    fontSize: stylesConstants.fontSize.big,
    fontWeight: 'bold',
    textAlign: 'center',
    paddingVertical: 10,
  },
  currentShelfContainer: {
    alignItems: 'center',
  },
  addToShelfContainer: {
    paddingHorizontal: stylesConstants.padding.normal,
  },
  addToShelfButton: {
    marginTop: 10,
  },
});
