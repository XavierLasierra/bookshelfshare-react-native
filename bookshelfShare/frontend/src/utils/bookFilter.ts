export function bookFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfListFilter(text: string, books: any) {
  return books.reduce((acc: any, book: any) => {
    if (book.bookData.title.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.publisher.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN13.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.isbn.ISBN10.toLowerCase().includes(text.toLowerCase())) return [...acc, book];
    if (book.bookData.authors.some((author: string) => author.toLowerCase()
      .includes(text.toLowerCase()))) return [...acc, book];

    return acc;
  }, []);
}

export function bookShelfLocationFilter(filterLocation: number[], books: any) {
  return books.filter(({ customInformation: { location } }: any) => JSON.stringify(location)
  === JSON.stringify(filterLocation));
}
