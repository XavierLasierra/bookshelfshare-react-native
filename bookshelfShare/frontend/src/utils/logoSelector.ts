import BookIcon from '../assets/bookIcon.svg';
import SearchIcon from '../assets/searchIcon.svg';
import ProfileIcon from '../assets/profileIcon.svg';

export default function logoSelector(logoName: string) {
  switch (logoName) {
    case 'SearchIcon':
      return SearchIcon;
    case 'BookIcon':
      return BookIcon;
    case 'ProfileIcon':
      return ProfileIcon;
    default:
      return BookIcon;
  }
}
