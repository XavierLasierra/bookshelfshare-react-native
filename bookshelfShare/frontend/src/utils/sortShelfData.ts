export default function sortShelfData(shelf: any) {
  return shelf?.books.reduce((acc: any, book: any) => {
    const foundShelfBox = acc.find(({ location }: any) => JSON.stringify(location)
         === JSON.stringify(book?.customInformation?.location));
    if (foundShelfBox) {
      foundShelfBox.books.push(book.bookData);
    } else {
      acc.push({
        location: book?.customInformation?.location,
        books: [book.bookData]
      });
    }
    return acc;
  }, []);
}
