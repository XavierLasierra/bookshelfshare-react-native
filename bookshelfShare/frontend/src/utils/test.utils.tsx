import React from 'react';
import { render as rtlRender } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import configureStore from '../redux/store';

function render(
  component: any,
  initialState: any
) {
  const store = configureStore(initialState);

  function Wrapper({ children }: any) {
    return (
      <Provider store={store}>
        {children}
      </Provider>
    );
  }
  return rtlRender(component, { wrapper: Wrapper });
}

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { render };
