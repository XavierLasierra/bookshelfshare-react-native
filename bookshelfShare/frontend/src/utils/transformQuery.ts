interface Query {
    isbn?: string,
    inauthor?: string,
    intitle?: string,
    inpublisher?: string
}

export default function transformQuery(query: Query):string {
  const queryEntries = Object.entries(query);
  const url = queryEntries
    .reduce((acc, queryElement, index) => `${acc}${queryElement[0]}=${queryElement[1]}${index < queryEntries.length - 1 ? '+' : ''}`, '');
  return `${url}&key=${process.env.GOOGLE_API_KEY}`;
}
