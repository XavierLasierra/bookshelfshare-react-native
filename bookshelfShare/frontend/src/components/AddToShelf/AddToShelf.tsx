import React, { useEffect, useState } from 'react';
import {
  View, SafeAreaView, Text, TouchableOpacity, ActivityIndicator
} from 'react-native';

import { useDispatch, useSelector } from 'react-redux';
import Header from '../Header/Header';
import ShelfSimulation from '../ShelfSimulation/ShelfSimulation';

import logoSelector from '../../utils/logoSelector';

import styles from './addToShelf.styles';
import globalStyles from '../../styles/global.styles';
import stylesConstants from '../../styles/styles.constants';
import { addToShelf } from '../../redux/actions/userShelves.creator';
import sortShelfData from '../../utils/sortShelfData';

export default function AddToShelf({
  navigation,
  route: {
    params: {
      shelf,
      logo,
      deleteFromShelf,
      notes,
      bookIsbn
    }
  }
}: any) {
  const dispatch = useDispatch();
  const { token, refreshToken } = useSelector((store: any) => store.tokens);
  const shelves = useSelector((store: any) => store.userShelves);
  const [isLoading, setIsLoading] = useState(false);
  const [location, setLocation] = useState([NaN, NaN]);
  const [invalidShelf, setInvalidShelf] = useState(false);

  function clickCallback(newLocation: number[]) {
    setInvalidShelf(false);
    setLocation(newLocation);
  }

  useEffect(() => {
    if (isLoading) {
      setIsLoading(false);
      navigation.pop();
    }
  }, [shelves]);

  function handleAddToShelf() {
    if (Number.isNaN(location[0]) || Number.isNaN(location[1])) {
      setInvalidShelf(true);
    } else {
      setIsLoading(true);
      dispatch(addToShelf(
        deleteFromShelf,
        // eslint-disable-next-line no-underscore-dangle
        shelf._id,
        bookIsbn, {
          notes,
          location
        },
        token,
        refreshToken
      ));
    }
  }

  return (
    <SafeAreaView>
      <Header Logo={logoSelector(logo)} BackButton navigation={navigation} />
      <View style={styles.addToShelfContainer}>
        <Text style={styles.addToShelfTitle}>Choose where to add it</Text>
        <View style={styles.currentShelfContainer}>
          <ShelfSimulation
            shelfSize={shelf.shelf}
            clickCallback={clickCallback}
            activeShelf={location}
            shelfData={sortShelfData(shelf)}
          />
        </View>
        {!isLoading
          ? (
            <TouchableOpacity
              style={[globalStyles.button, styles.addToShelfButton]}
              testID="addButton"
              onPress={handleAddToShelf}
            >
              <Text style={globalStyles.buttonText}>Add to shelf</Text>
            </TouchableOpacity>
          )
          : (
            <TouchableOpacity
              style={[globalStyles.button, styles.addToShelfButton]}
              testID="addButton"
              disabled
            >
              <ActivityIndicator testID="loadingIndicator" size="large" color={stylesConstants.colors.white} />
            </TouchableOpacity>
          )}
        {invalidShelf && <Text style={globalStyles.invalid}>Select a shelf to add it</Text>}
      </View>
    </SafeAreaView>
  );
}
