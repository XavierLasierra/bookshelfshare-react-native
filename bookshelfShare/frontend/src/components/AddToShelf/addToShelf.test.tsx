import React from 'react';
import AddToShelf from './AddToShelf';
import { render, fireEvent } from '../../utils/test.utils';
import shelfMock from '../../mocks/shelf.mock';
import { addToShelf } from '../../redux/actions/userShelves.creator';
import userShelvesActions from '../../redux/actions/userShelves.actions';

const navigation = {
  pop: jest.fn()
};

const route = {
  params: {
    shelf: shelfMock
  }
};

jest.mock('../../redux/actions/userShelves.creator');

describe('Given an AddToShelf component', () => {
  describe('When it is rendered', () => {
    describe('And isbn is Not found', () => {
      let screen: any;
      beforeEach(() => {
        (addToShelf as jest.Mock).mockReturnValue({ type: '' });

        screen = render(
          <AddToShelf navigation={navigation} route={route} />
        );
      });
      test('Then should match the snapshot', () => {
        expect(screen).toMatchSnapshot();
      });

      describe('And addButton is clicked without selecting any shelf', () => {
        test('Then should render a Text with testId invalidShelf', () => {
          const addButton = screen.getByTestId('addButton');
          fireEvent.press(addButton);

          expect(screen.queryByTestId('invalidShelf')).not.toBe(null);
        });
      });

      describe('And addButton is clicked after selecting one shelf', () => {
        describe('And the updated shelf is loading', () => {
          beforeEach(() => {
            const shelfButton = screen.getByTestId('shelf-0-0');
            const addButton = screen.getByTestId('addButton');
            fireEvent.press(shelfButton);
            fireEvent.press(addButton);
          });

          test('Then shoud call addToShelf function', () => {
            expect(addToShelf).toHaveBeenCalled();
          });

          test('Then shoud render an ActivityIndicator with testID loadingIndicator', () => {
            expect(screen.queryByTestId('loadingIndicator')).not.toBe(null);
          });
        });

        describe('And the updated shelf is resolved', () => {
          test('Then should call navigation.pop', () => {
            (addToShelf as jest.Mock).mockReturnValue({
              type: userShelvesActions.ADD_BOOK_TO_SHELF,
              data: shelfMock
            });

            screen = render(
              <AddToShelf navigation={navigation} route={route} />
            );

            const shelfButton = screen.getByTestId('shelf-0-0');
            const addButton = screen.getByTestId('addButton');
            fireEvent.press(shelfButton);
            fireEvent.press(addButton);

            expect(navigation.pop).toHaveBeenCalled();
          });
        });
      });
    });
  });
});
