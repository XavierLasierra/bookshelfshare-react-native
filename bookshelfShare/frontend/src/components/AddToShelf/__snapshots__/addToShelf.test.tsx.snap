// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an AddToShelf component When it is rendered And isbn is Not found Then should match the snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "left": 20,
        "opacity": 1,
        "position": "absolute",
        "top": 20,
        "zIndex": 1,
      }
    }
    testID="backButton"
  >
    <IconMock
      height={30}
      style={
        Object {
          "transform": Array [
            Object {
              "rotateZ": "180deg",
            },
          ],
        }
      }
      width={30}
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "backgroundColor": "#F5F5F5",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#D9D8C4",
          "borderBottomLeftRadius": 25,
          "justifyContent": "center",
          "padding": 10,
          "width": "100%",
        }
      }
    >
      <IconMock
        height={45}
        width={45}
      />
    </View>
    <View
      style={
        Object {
          "height": 0,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#D9D8C4",
            "height": 20,
            "width": 20,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#F5F5F5",
              "borderTopRightRadius": 25,
              "height": 20,
              "width": 20,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "paddingHorizontal": 30,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
          "paddingVertical": 10,
          "textAlign": "center",
        }
      }
    >
      Choose where to add it
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
        }
      }
    >
      <View
        style={
          Object {
            "height": 450,
            "justifyContent": "center",
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#958472",
              "borderColor": "#2F2406",
              "borderRadius": 15,
              "borderTopWidth": 10,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderColor": "#958472",
                  "borderWidth": 6,
                  "elevation": 10,
                },
                Object {
                  "height": 400,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                  },
                  Object {
                    "height": "50%",
                    "width": "100%",
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#958472",
                      "height": "100%",
                      "justifyContent": "center",
                    },
                    Object {
                      "width": "50%",
                    },
                  ]
                }
              >
                <View
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#4D3C2A",
                      "borderBottomWidth": 5,
                      "borderColor": "#2F2406",
                      "borderRightWidth": 5,
                      "height": "90%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": "90%",
                    }
                  }
                  testID="shelf-0-0"
                >
                  <Text
                    style={
                      Object {
                        "color": "#FFFFFF",
                        "fontSize": 16,
                        "fontWeight": "bold",
                      }
                    }
                  >
                    2
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "#FFFFFF",
                        "fontSize": 16,
                        "fontWeight": "bold",
                      }
                    }
                  >
                    books
                  </Text>
                </View>
              </View>
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#958472",
                      "height": "100%",
                      "justifyContent": "center",
                    },
                    Object {
                      "width": "50%",
                    },
                  ]
                }
              >
                <View
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#4D3C2A",
                      "borderBottomWidth": 5,
                      "borderColor": "#2F2406",
                      "borderRightWidth": 5,
                      "height": "90%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": "90%",
                    }
                  }
                  testID="shelf-0-1"
                >
                  <Text
                    style={
                      Object {
                        "color": "#FFFFFF",
                        "fontSize": 16,
                        "fontWeight": "bold",
                      }
                    }
                  >
                    0
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "#FFFFFF",
                        "fontSize": 16,
                        "fontWeight": "bold",
                      }
                    }
                  >
                    books
                  </Text>
                </View>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                  },
                  Object {
                    "height": "50%",
                    "width": "100%",
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#958472",
                      "height": "100%",
                      "justifyContent": "center",
                    },
                    Object {
                      "width": "50%",
                    },
                  ]
                }
              >
                <View
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#4D3C2A",
                      "borderBottomWidth": 5,
                      "borderColor": "#2F2406",
                      "borderRightWidth": 5,
                      "height": "90%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": "90%",
                    }
                  }
                  testID="shelf-1-0"
                >
                  <Text
                    style={
                      Object {
                        "color": "#FFFFFF",
                        "fontSize": 16,
                        "fontWeight": "bold",
                      }
                    }
                  >
                    0
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "#FFFFFF",
                        "fontSize": 16,
                        "fontWeight": "bold",
                      }
                    }
                  >
                    books
                  </Text>
                </View>
              </View>
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#958472",
                      "height": "100%",
                      "justifyContent": "center",
                    },
                    Object {
                      "width": "50%",
                    },
                  ]
                }
              >
                <View
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  nativeID="animatedComponent"
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#4D3C2A",
                      "borderBottomWidth": 5,
                      "borderColor": "#2F2406",
                      "borderRightWidth": 5,
                      "height": "90%",
                      "justifyContent": "center",
                      "opacity": 1,
                      "width": "90%",
                    }
                  }
                  testID="shelf-1-1"
                >
                  <Text
                    style={
                      Object {
                        "color": "#FFFFFF",
                        "fontSize": 16,
                        "fontWeight": "bold",
                      }
                    }
                  >
                    0
                  </Text>
                  <Text
                    style={
                      Object {
                        "color": "#FFFFFF",
                        "fontSize": 16,
                        "fontWeight": "bold",
                      }
                    }
                  >
                    books
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#000000",
          "borderRadius": 10,
          "display": "flex",
          "height": 70,
          "justifyContent": "center",
          "marginTop": 10,
          "opacity": 1,
          "width": "100%",
        }
      }
      testID="addButton"
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 20,
            "fontWeight": "500",
          }
        }
      >
        Add to shelf
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
